### PR TITLE
PYTHON-2723 Convert transactions spec tests to unified test format

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
The tests are taking a long time because many are waiting 60 seconds for open transactions to time out. We need to add killAllSessions as described here: https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.md#terminating-open-transactions

Note we already do this in the legacy txn test runner here:
https://github.com/mongodb/mongo-python-driver/blob/992d1507e7a49292b555cc0201d95d49234c6228/test/utils_spec_runner.py#L177-L185

**Comment by blink1073** [comment]
evergreen retry

**Comment by blink1073** [comment]
evergreen retry

**Comment by blink1073** [comment]
Evergreen is failing to create the patch build because the diff is too big, here is a manual patch: https://spruce.mongodb.com/version/66022b707612660007a780f0/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Comment by blink1073** [comment]
I verified that the transaction tests are run on serverless.

File: test/unified_format.py [hunk 1]
```diff
@@ -1229,6 +1235,16 @@ def _databaseOperation_createCommandCursor(self, target, **kwargs):
 
         return cursor
 
+    def kill_all_sessions(self):
+        if getattr(self, "client", None) is None:
```
Comment (ShaneHarvey) [inline test/unified_format.py:1239]: Why would client be None here? Should we be using the client_context.client?

File: test/unified_format.py [hunk 2]
```diff
@@ -1229,6 +1235,16 @@ def _databaseOperation_createCommandCursor(self, target, **kwargs):
...
+            return
+        try:
+            self.client.admin.command("killAllSessions", [])
```
Comment (ShaneHarvey) [inline test/unified_format.py:1242]: In the legacy runner, we would run this on all mongoses. I think we should do that here too.

File: test/unified_format.py [hunk 3]
```diff
@@ -1699,6 +1719,15 @@ def run_operations(self, spec):
...
                 self.run_entity_operation(op)
 
+    def run_operations_and_throw(self, spec):
+        for op in spec:
+            if op["object"] == "testRunner":
+                self.run_special_operation(op)
+            else:
+                result = self.run_entity_operation(op)
+                if isinstance(result, Exception):
+                    raise result
```
Comment (ShaneHarvey) [inline test/unified_format.py:1729]: Why re-raise this exception after ignoreResultAndError or expectError? Does the spec say to do that somewhere?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/unified_format.py (hunk 1 above)
Comment (blink1073) [reply test/unified_format.py:1239]: It was None on two of the tests that were testing the spec runner itself without creating a client.

File: test/unified_format.py (hunk 3 above)
Comment (blink1073) [reply test/unified_format.py:1729]: Yes, "Ensure that neither ignoreResultAndError nor expectError interfere with propagating exceptions from a callback operation to withTransaction"

File: test/unified_format.py (hunk 2 above)
Comment (blink1073) [reply test/unified_format.py:1242]: I couldn't find the equivalent of `mongos_clients`, it looks like all clients go through this line: https://github.com/mongodb/mongo-python-driver/blob/f757fe39cc12520e299090adf750fc07eb03dec8/test/unified_format.py#L501

File: test/unified_format.py (hunk 3 above)
Comment (ShaneHarvey) [reply test/unified_format.py:1729]: Where is that written? That line doesn't appear in the spec repo.

File: test/unified_format.py (hunk 2 above)
Comment (ShaneHarvey) [reply test/unified_format.py:1242]: We'll need to create those clients like is done in the legacy txn tests: https://github.com/mongodb/mongo-python-driver/blob/992d1507e7a49292b555cc0201d95d49234c6228/test/test_transactions.py#L71

File: test/unified_format.py (hunk 1 above)
Comment (ShaneHarvey) [reply test/unified_format.py:1239]: Oh I see it's TestCreateEntities.

File: test/unified_format.py (hunk 3 above)
Comment (blink1073) [reply test/unified_format.py:1729]: That is from the ticket description, there is more discussion on the original spec PR.

File: test/unified_format.py (hunk 2 above)
Comment (blink1073) [reply test/unified_format.py]: Done

**Comment by blink1073** [comment]
New manual patch: https://spruce.mongodb.com/version/660329f0ca19fe0007f917c6/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Comment by ShaneHarvey** [comment]
What's up with the "@evergreen-ci-prod evergreen — Evergreen error" in the check?

**Comment by ShaneHarvey** [comment]
```
 [2024/03/26 13:15:00.064] ERROR: failed on setup with "pymongo.errors.ConfigurationError: Cannot specify directConnection=true with loadBalanced=true" ()
 [2024/03/26 13:15:00.064] cls = <class 'test.test_auth_spec.TestUnifiedReauthenticateWithRetry'>
 [2024/03/26 13:15:00.064]     @classmethod
 [2024/03/26 13:15:00.064]     def setUpClass(cls):
 [2024/03/26 13:15:00.064]         # super call creates internal client cls.client
 [2024/03/26 13:15:00.064]         super().setUpClass()
 [2024/03/26 13:15:00.064]         # process file-level runOnRequirements
 [2024/03/26 13:15:00.064]         run_on_spec = cls.TEST_SPEC.get("runOnRequirements", [])
 [2024/03/26 13:15:00.064]         if not cls.should_run_on(run_on_spec):
 [2024/03/26 13:15:00.064]             raise unittest.SkipTest(f"{cls.__name__} runOnRequirements not satisfied")
 [2024/03/26 13:15:00.064]     
 [2024/03/26 13:15:00.064]         # Handle mongos_clients for transactions tests.
 [2024/03/26 13:15:00.064]         cls.mongos_clients = []
 [2024/03/26 13:15:00.064]         if client_context.supports_transactions():
 [2024/03/26 13:15:00.064]             for address in client_context.mongoses:
 [2024/03/26 13:15:00.064] >               cls.mongos_clients.append(single_client("{}:{}".format(*address)))
 [2024/03/26 13:15:00.064] test/unified_format.py:1022: 
 [2024/03/26 13:15:00.064] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 [2024/03/26 13:15:00.064] test/utils.py:604: in single_client
 [2024/03/26 13:15:00.064]     return _mongo_client(h, p, directConnection=True, **kwargs)
 [2024/03/26 13:15:00.064] test/utils.py:594: in _mongo_client
 [2024/03/26 13:15:00.064]     return MongoClient(uri, port, **client_options)
 [2024/03/26 13:15:00.064] pymongo/mongo_client.py:825: in __init__
 [2024/03/26 13:15:00.064]     _check_options(seeds, opts)
 [2024/03/26 13:15:00.064] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 [2024/03/26 13:15:00.064] nodes = {('32071-drivertest-lb.3x79d.mongodb-dev.net', 27017)}
 [2024/03/26 13:15:00.064] options = {'loadBalanced': True, 'compressors': ['zlib'], 'tls': True, 'directConnection': True, 'username': 'drivers-testing', 'password': '47l0tpPl13dobvHF', 'document_class': <class 'dict'>, 'tz_aware': False, 'connect': True}
 [2024/03/26 13:15:00.064]     def _check_options(nodes: Sized, options: Mapping[str, Any]) -> None:
 [2024/03/26 13:15:00.064]         # Ensure directConnection was not True if there are multiple seeds.
 [2024/03/26 13:15:00.064]         if len(nodes) > 1 and options.get("directconnection"):
 [2024/03/26 13:15:00.064]             raise ConfigurationError("Cannot specify multiple hosts with directConnection=true")
 [2024/03/26 13:15:00.064]     
 [2024/03/26 13:15:00.064]         if options.get("loadbalanced"):
 [2024/03/26 13:15:00.064]             if len(nodes) > 1:
 [2024/03/26 13:15:00.064]                 raise ConfigurationError("Cannot specify multiple hosts with loadBalanced=true")
 [2024/03/26 13:15:00.064]             if options.get("directconnection"):
 [2024/03/26 13:15:00.064] >               raise ConfigurationError("Cannot specify directConnection=true with loadBalanced=true")
 [2024/03/26 13:15:00.064] E               pymongo.errors.ConfigurationError: Cannot specify directConnection=true with loadBalanced=true
 [2024/03/26 13:15:00.064] pymongo/uri_parser.py:403: ConfigurationError
```

Looks like this needs to be skipped on serverless+loadBalanced clusters.

**Comment by blink1073** [comment]
> What's up with the "@evergreen-ci-prod evergreen — Evergreen error" in the check?

The diff is too big, they're working on a fix where it will at least submit the patch even if the diff part fails.

**Comment by blink1073** [comment]
> Looks like the mongos_clients logic needs to be skipped on serverless+loadBalanced clusters.
Done

https://spruce.mongodb.com/version/66033b7a8b69b6000726fb32/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Review by ShaneHarvey — APPROVED** [review]
LGTM

**Pull request merged**